File: Java/app/src/main/java/com/tyhoo/mvvm/java/ui/MainViewModel.java
package com.tyhoo.mvvm.java.ui;

import android.content.Context;
import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.google.common.base.Strings;
import com.google.gson.Gson;
import com.tyhoo.mvvm.java.R;
import com.tyhoo.mvvm.java.data.User;

public class MainViewModel extends ViewModel {

    private static final String TAG = MainViewModel.class.getSimpleName();

    private final MutableLiveData<Boolean> mDataLoading = new MutableLiveData<>();

    private final MutableLiveData<String> mUserData = new MutableLiveData<>();

    private String jsonResponse;

    public void start(Context context) {
        networkRequest(context);
    }

    public void parseJson(Context context) {
        if (Strings.isNullOrEmpty(jsonResponse)) {
            return;
        }

        Gson gson = new Gson();
        User user = gson.fromJson(jsonResponse, User.class);
        String text = context.getString(R.string.parse_text, user.getLogin(), user.getHtmlUrl(),
                user.getName(), user.getCompany(), user.getLocation());
        mUserData.setValue(text);
    }

    private void networkRequest(Context context) {
        mDataLoading.setValue(true);

        RequestQueue queue = Volley.newRequestQueue(context);
        String url = "https://api.github.com/users/cnwutianhao";

        StringRequest stringRequest = new StringRequest(Request.Method.GET, url,
                response -> {
                    Log.d(TAG, "networkRequest: success");
                    jsonResponse = response;
                    mDataLoading.setValue(false);
                    mUserData.setValue(response);
                },
                error -> {
                    Log.d(TAG, "networkRequest: error");
                    mDataLoading.setValue(false);
                    mUserData.setValue("");
                });

        queue.add(stringRequest);
    }

    public LiveData<Boolean> getDataLoading() {
        return mDataLoading;
    }

    public MutableLiveData<String> getUserData() {
        return mUserData;
    }
}

File: Kotlin/app/src/main/java/com/tyhoo/mvvm/kotlin/ui/MainViewModel.kt
package com.tyhoo.mvvm.kotlin.ui

import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.android.volley.Request
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.google.gson.Gson
import com.tyhoo.mvvm.kotlin.R
import com.tyhoo.mvvm.kotlin.data.User

class MainViewModel : ViewModel() {

    private val _dataLoading = MutableLiveData<Boolean>()
    val dataLoading: LiveData<Boolean>
        get() = _dataLoading

    private val _userData = MutableLiveData<String>()
    val userData: LiveData<String>
        get() = _userData

    private var jsonResponse = ""

    fun start(context: Context) {
        networkRequest(context)
    }

    private fun networkRequest(context: Context) {
        _dataLoading.value = true

        val queue = Volley.newRequestQueue(context)
        val url = "https://api.github.com/users/cnwutianhao"

        val stringRequest = StringRequest(Request.Method.GET, url,
            { response ->
                jsonResponse = response
                _dataLoading.value = false
                _userData.value = response
            }) {
            _dataLoading.value = false
            _userData.value = ""
        }

        queue.add(stringRequest)
    }

    fun parseJson(context: Context) {
        if (jsonResponse.isEmpty()) {
            return
        }

        val gson = Gson()
        val user = gson.fromJson(jsonResponse, User::class.java)
        val text = context.getString(
            R.string.parse_text, user.login, user.html_url,
            user.name, user.company, user.location
        )

        _userData.value = text
    }
}